rgw/v2/tests/s3_swift/test_object_level_retention.py

```diff
@@ -0,0 +1,183 @@
+password = "32characterslongpassphraseneeded".encode("utf-8")
+encryption_key = hashlib.md5(password).hexdigest()
+
+
+def test_exec(config, ssh_con):
+    io_info_initialize = IOInfoInitialize()
+    basic_io_structure = BasicIOInfoStructure()
+    io_info_initialize.initialize(basic_io_structure.initial())
+    ceph_conf = CephConfOp(ssh_con)
+    rgw_service = RGWService()
+
+    # create user
+    all_users_info = s3lib.create_users(config.user_count)
+    for each_user in all_users_info:
+        # authentication
+        auth = Auth(each_user, ssh_con, ssl=config.ssl)
+        s3_conn_client = auth.do_auth_using_client()
+        # create buckets with object lock configuration
+        if config.test_ops["create_bucket"] is True:
+            log.info(f"no of buckets to create: {config.bucket_count}")
+            for bc in range(config.bucket_count):
+                bucket_name_to_create = utils.gen_bucket_name_from_userid(
+                    each_user["user_id"], rand_no=bc
+                )
+                log.info(f"creating bucket with name: {bucket_name_to_create}")
+                rgw_ip_and_port = get_rgw_ip_and_port(ssh_con)
+                s3_conn_client.create_bucket(
+                    Bucket=bucket_name_to_create, ObjectLockEnabledForBucket=True
+                )
+                if config.test_ops["create_object"] is True:
+                    # uploading data in mentioned mode
+                    if config.test_ops["object_compliance"]:
+                        lock_mode = "COMPLIANCE"
+                    else:
+                        lock_mode = "GOVERNANCE"
+
+                    # defining retain until date for longer
+                    retain_until = "2025-01-01T00:00:00"
+                    log.info(f"s3 objects to create: {config.objects_count}")
+                    for oc, size in list(config.mapped_sizes.items()):
+                        s3_object_name = utils.gen_s3_object_name(
+                            bucket_name_to_create, 0
+                        )
+                        log.info(f"s3 object name: {s3_object_name}")
+                        s3_object_path = os.path.join(TEST_DATA_PATH, s3_object_name)
+                        log.info(f"s3 object path: {s3_object_path}")
+                        log.info(f"upload type: normal with lock mode {lock_mode}")
+                        io_generator(TEST_DATA_PATH + "/" + s3_object_name, size)
+                        obj_body = TEST_DATA_PATH + "/" + s3_object_name
+                        reusable.object_lock_retention(
+                            s3_conn_client,
+                            bucket_name_to_create,
+                            s3_object_name,
+                            obj_body,
+                            lock_mode,
+                            retain_until,
+                        )
+
+                log.info("Try to delete the uploaded object versions")
+                versions = s3_conn_client.list_object_versions(
```

Comment:
where do we handled uploading of version objects?

Reply:
When we enable locking on bucket , the bucket becomes versioned by default

Reply:
> When we enable locking on bucket , the bucket becomes versioned by default

As per current conf we are uploading only single version of object ryt..Can we have a conf for multiple version upload and test this scenario?

Reply:
actually as per current conf we are uploading single key and 30 versions of the same key